src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +30,371 @@
+        final Object[] buildingArray = buildingList.toArray();
+
+        final int random = worker.getRandom().nextInt(buildingArray.length);
+        final AbstractBuilding building = (AbstractBuilding) buildingArray[random];
+
+        if (building instanceof BuildingWareHouse
+                || building instanceof BuildingDeliveryman || building instanceof BuildingTownHall)
+        {
+            return null;
+        }
+
+        return building.getLocation();
+    }
+
+    /**
+     * Dump the inventory into the warehouse.
+     * @return the next state to go to.
+     */
+    public AIState dump()
+    {
+        if(!worker.isWorkerAtSiteWithMove(wareHouse.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return DUMPING;
+        }
+
+        wareHouse.getTileEntity().dumpInventoryIntoWareHouse(worker.getInventoryCitizen());
+        gatherTarget = null;
+
+        return START_WORKING;
+    }
+
+
+    /**
+     * Deliver the items to the hut.
+     * @return the next state.
+     */
+    private AIState deliver()
+    {
+        final AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            final AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if(!worker.isWorkerAtSiteWithMove(buildingToDeliver.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+                {
+                    return DELIVERY;
+                }
+
+                final InventoryCitizen workerInventory  = worker.getInventoryCitizen();
+                for(int i = 0; i < workerInventory.getSizeInventory(); i++)
+                {
+                   final  ItemStack stack = workerInventory.getStackInSlot(i);
+                    if(stack == null)
+                    {
+                        continue;
+                    }
+
+                    if(InventoryUtils.addItemStackToInventory(buildingToDeliver.getTileEntity(), stack))
+                    {
+                        workerInventory.removeStackFromSlot(i);
+                    }
+                }
+                buildingToDeliver.setOnGoingDelivery(false);
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Prepare deliveryman for delivery.
+     * Check if the building still needs the item and if the required items are still in the warehouse.
+     * @return the next state to go to.
+     */
+    private AIState prepareDelivery()
+    {
+        final AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            final AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if(buildingToDeliver != null)
+            {
+                boolean ableToDeliver = wareHouse.getTileEntity().checkInWareHouse(buildingToDeliver);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'ableToDeliver' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)